src/Contract/Transaction.purs

```diff
@@ -201,13 +211,171 @@ calculateMinFeeM
   :: forall (r :: Row Type). Transaction -> Contract r (Maybe Coin)
 calculateMinFeeM = map hush <<< calculateMinFee
 
+-- | Helper to adapt to UsedTxOuts
+withUsedTxouts
+  :: forall (r :: Row Type) (a :: Type)
+   . ReaderT UsedTxOuts (Contract r) a
+  -> Contract r a
+withUsedTxouts f = asks (_.usedTxOuts <<< unwrap) >>= runReaderT f
+
 -- | Attempts to balance an `UnattachedUnbalancedTx`.
 balanceTx
   :: forall (r :: Row Type)
    . UnattachedUnbalancedTx
   -> Contract r (Either BalanceTxError.BalanceTxError UnattachedTransaction)
 balanceTx = wrapContract <<< BalanceTx.balanceTx
 
+-- Helper to avoid repetition
+withTransactions
+  :: forall (a :: Type)
+       (t :: Type -> Type)
+       (r :: Row Type)
+       (tx :: Type)
+   . Traversable t
+  => (t UnattachedUnbalancedTx -> Contract r (t tx))
+  -> (tx -> Transaction)
+  -> t UnattachedUnbalancedTx
+  -> (t tx -> Contract r a)
+  -> Contract r a
+withTransactions prepare extract utxs action = do
+  txs <- prepare utxs
+  res <- try $ action txs
+  _ <- traverse (\t -> withUsedTxouts $ unlockTransactionInputs t) $ map extract
+    $
+      txs
+  case res of
+    Left e -> throwError e
+    Right x -> pure x
+
+withSingleTransaction
+  :: forall (a :: Type) (tx :: Type) (r :: Row Type)
+   . (UnattachedUnbalancedTx -> Contract r tx)
+  -> (tx -> Transaction)
+  -> UnattachedUnbalancedTx
+  -> (tx -> Contract r a)
+  -> Contract r a
+withSingleTransaction prepare extract utx action =
+  withTransactions (traverse prepare) extract (NonEmptyArray.singleton utx)
+    (action <<< NonEmptyArray.head)
+
+-- | Execute an action on an array of balanced
+-- | transactions (balanceTxs will be called). Within
+-- | this function, all transaction inputs used by these
+-- | transactions will be locked, so that they are not used
+-- | in any other context.
+-- | After the function completes, the locks will be removed.
+-- | Errors will be thrown.
+withBalancedTxs
+  :: forall (a :: Type)
+       (r :: Row Type)
+   . (Array UnattachedUnbalancedTx)
+  -> (Array UnattachedTransaction -> Contract r a)
+  -> Contract r a
+withBalancedTxs = withTransactions balanceTxs get1
+
+-- | Execute an action on a balanced transaction (balanceTx will
+-- | be called). Within this function, all transaction inputs
+-- | used by this transaction will be locked, so that they are not
+-- | used in any other context.
+-- | After the function completes, the locks will be removed.
+-- | Errors will be thrown.
+withBalancedTx
+  :: forall (a :: Type)
+       (r :: Row Type)
+   . UnattachedUnbalancedTx
+  -> (UnattachedTransaction -> Contract r a)
+  -> Contract r a
+withBalancedTx = withSingleTransaction balanceTx' get1
+  where
+  balanceTx' uutx = do
+    tx <- balanceTx uutx
```

Comment:
Again, stylistic thing here, but the way we do it in most of the codebase is 

```suggestion
    balanceTx uutx >>= case _ of
```

but I'd even recommend just using `Contract.Monad.liftedE`